### PR TITLE
WIP - Refactor & and add snapshot configuration

### PR DESCRIPTION
**Pull request by hepiyellow**
# This PR is on HOLD !
Waiting for separate PR which only migrates to yoshi

## Refactor
Motivation: 
- complex code... divide into files. Make it easier to understand.
- Allow easier (manual) testing. And easier unit-testing, if we decide to write tests for this (:

#### Divide code into:
- `eyes.js` - initializes eyes and exports a singleton.
- `eyes-it.js` - The actual `it` method that will run. Accepts a configuration object.
- `eyes-hook.js` - "connects" the `eyes-it` function on the `eyes` singleton. Takes care of converting `it` 3rd argument to `eyes-it`  config object.

#### Hooking & Mocking
- Separate "hooking" of `eyes.it` from the mocking of `eyes` functions that don't work when there is no API key set. This is useful when unit-testing `eyes-it` without an API key.

### Changes in `eyes-it` logic
- add `enableSnapshotAtBrowserGet` and `enableSnapshotAtEnd` to eyes.it configuration (3rd argument)
- Split "after" hook, to first take a snapshot, and to the `eyes.close` only after the last "after function". (To allow consumer to take a snapshot  in an `afterEach` function.

### Add yoshi
- [ ] release is still using `wnpm-release` we might want to change to `yoshi release`.
- [ ] need to test if `wnpm-release` is still working (after I put code under `src` and added `yoshi` build to `dist`)

### Tests
Add "e2e" test, which is not really a true integration test, since I am mocking `eyes` and also mocking `browser.get`.

### CI
- [ ] Why is pull-request CI not working ?

### CODE REVIEW

File: eyes.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    batchName = null;
+    batchStartAt = 0;
+  }
+
+  if (process.env.EYES_API_KEY) {
+    eyes.setApiKey(process.env.EYES_API_KEY);
+  } else {
+    // mock eyes function which throw error when Api key not set
+    eyes.checkWindow = () => Promise.resolve();
+    eyes.checkRegionBy = () => Promise.resolve();
```
Comment (hepiyellow) [inline eyes.js]: Added here more mock functions

File: eyes-it.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  return 'eyes.it ' + spec.getFullName() + versionDescription;
+}
+
+function handleError(err, done) {
+  fail(err);
+  done();
+}
+
+var _enableSnapshotAtBrowserGet = false;
```
Comment (hepiyellow) [inline eyes-it.js]: This replaces the old `eyesOpen`

File: eyes-it.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  spec.beforeAndAfterFns = function () {
+    //TODO: are we sure that we need to pass args (msg, itFn) to hooked?
+    var result = hooked.apply(this, _itArgs);
+    result.befores.unshift({
+      fn: function (done) {
+        _enableSnapshotAtBrowserGet = enableSnapshotAtBrowserGet;
+        eyes.open(browser, appName, buildSpecName(spec, specVersion), windowSize).then(done);
+      }, timeout: () => 30000
+    });
+    result.afters.push(
```
Comment (hepiyellow) [inline eyes-it.js]: This used to be `result.afters.ushift`

Comment (shlomitc) [reply eyes-it.js]: This is a breaking change. In many cases we reset the state of the page after the test. So This now means that the final screenshot will be different

Comment (hepiyellow) [reply eyes-it.js]: Good point !
We can split the `afters` hooks , and put the `checkWindow` at the beginning,
and the `eyes.close` at the end.
That way everybody wins.

**Comment by hepiyellow** [comment]
@shlomitc I was very concerned with missing tests. And still thinking about it.
It's a bit tricky to test e2e, since you need a testsuit inside a test suit.
I've see some yoshi tests which do that.
 but I thought of a simple trick, tell me what you think.

- We can have 2 test phases.
- In the first one we will run real test suits which use eyes.it exactly as a consumer does. This phase will produce files (whenever checkWindow is called) in a temp folder.
- In the 2nd phase  (2nd test suit) we will simply assert on the produced files.

**Comment by hepiyellow** [comment]
I found a simpler way to test it.
The assumption I am making is that `eyes.it` operates only in the scope of an `it` block. So the `beforeAll` and `afterAll` are not affected by `eyes.it`.
That allows me to use them to test what happened in the `it` block (without having 2 phases).

Here is a full example:
```js
import eyes from '../src';
import { hookEyesIt } from '../src/eyes-hook';

hookEyesIt();
let logs = [];
function log(msg) {
  logs.push(msg);
}

/** Mocks */
eyes.open = (browser, appName, specName, windowSize) => {
  log(`open(${specName})`);
  return Promise.resolve();
};

eyes.checkWindow = msg => {
  log(`checkWindow(${msg}}`);
  return Promise.resolve();
};

export class Test {
  _it = () => null;
  logMatchers = [];

  it(_it) {
    this._it = _it;
    return this;
  }

  expect(logMatchers) {
    this.logMatchers = logMatchers;
    return this;
  }

  run() {
    describe('', () => {
      this._it();

      afterAll(() => {
        expect(this.logMatchers.length).toEqual(logs.length); // eslint-disable-line jest/prefer-to-have-length
        this.logMatchers.forEach((matcher, index) => {
          expect(logs[index]).toEqual(jasmine.stringMatching(matcher));
        });
        logs = [];
      });
    });
  }
}

describe('eyes.it', () => {
  new Test()
    .it(() => {
      eyes.it('should create snapshot at end', async () => {});
    })
    .expect([/^open/, /^checkWindow/])
    .run();

  new Test()
    .it(() => {
      eyes.it('should NOT create snapshot at end', async () => {}, {
        enableSnapshotAtEnd: false,
      });
    })
    .expect([/^open/])
    .run();
});

```

**Comment by yanivefraim** [comment]
I think we must have at least two e2e/it tests:

1. Have a real `eyes.it` test which, when passes, logs 'success'. We can assert for this log (the log is in production code, maybe only for debug mode. This is the only way imo to know that eyes succeed.
2. Have a test which supposed to fail (we change something in test's "production" code so test should fail)

An example for something similar can be found here: https://github.com/wix-incubator/match-screenshot

**Comment by hepiyellow** [comment]
@yanivefraim 
Well, It's true that without an integration test, this library can not be fully tested.
I just think it is a complicated test which brings little value.

The `eyes.it` library is mostly about adding logic (hooks) to the test runner and protractor's `browser.get`, and it less about `eyes` itself.
I would feel pretty comfortable with a simple unit-test that mocks `eyes`.

> The only `eyes` behaviour we need to mock is that `checkWindow` can not be called without calling `eyes.open` first. And handling an `eyes` error.

But assuming we are adding integration test...
I saw in your example that with puppeteer, you can [set the page content](https://github.com/wix-incubator/match-screenshot/blob/082b12cba856845efbce716fba5f73995433b858/tests/__fixtures__/test.mocha.js#L16).
I guess that with protractor it is not possible, so I'd have to have a static page, and little web-server to serve it. Right?

**Pull request closed without merging**